### PR TITLE
Fix unifying undefined first class function types issue.

### PR DESCRIPTION
**Pull request by pearu**
Tackles #5615 .

### CODE REVIEW

File: numba/tests/test_function_type.py [hunk 1]
```diff
@@ -1072,3 +1072,28 @@ def bar_good(foos):
...
+    def test_issue_5615(self):
+        @njit
+        def foo1(x):
+            return x + 1
+
+        @njit
+        def foo2(x):
+            return x + 2
+
+        from numba import literal_unroll
```
Comment (stuartarchibald) [inline numba/tests/test_function_type.py:1085]: Think this needs to be a global (there's a bug, don't recall ticket, which makes `literal_unroll` as a freevar not work).

Comment (stuartarchibald) [reply numba/tests/test_function_type.py:1085]: Fixed the bug https://github.com/numba/numba/pull/5638

Comment (pearu) [reply numba/tests/test_function_type.py:1085]: Thanks @stuartarchibald, I will update this PR accordingly.

Comment (pearu) [reply numba/tests/test_function_type.py]: Done.

File: numba/core/utils.py [hunk 2]
```diff
@@ -532,8 +532,6 @@ def unified_function_type(numba_types, require_precise=True):
             if isinstance(t, types.UndefinedFunctionType):
                 if undefined_function is None:
                     undefined_function = t
-                else:
-                    assert undefined_function == t
```
Comment (sklam) [inline numba/core/utils.py:536]: I'm trying to understand this patch. I see that when this condition is reached, this function will return None. Couldn't it return `None` at this point?

**Comment by stuartarchibald** [comment]
Indirectly merged by https://github.com/numba/numba/pull/5658 thanks for the patch.

**Pull request merged**